### PR TITLE
Add item type in front of item fields

### PR DESCRIPTION
**Pull request by 0xNath**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14688

### CODE REVIEW

**Comment by cconard96** [comment]
Hello @0xNath, thank you for your contribution.

I tested your fix, but it doesn't seem to add any prefixes to the items in the display preference list.
Looking at the code, I am not sure about the approach.

In a search option, the `name` property is the label for that field/option rather than the "group".
In the Search engine, this code here determines the group name (if any) for prefixing the column headers:
https://github.com/glpi-project/glpi/blob/8fbda207a9ae92757078492b19fcb45dfc675fa0/src/Search.php#L1576-L1603

It is a little complex, but the process for each column is:
1. While key for the current search option entry isn't null (haven't reached the end of the options array) and isn't the ID for the column/search option we are looking for, the pointer is moved to the next element
2. Once the option is found, it enumerates the options backwards until it finds one that has a non-numeric key. These non-numeric key entries are group labels.
3. If the key isn't "common" and the column data doesn't already have a group name, the value for the found element is set as the group name.
4. Reset the pointer for the search options array so we start looking from the beginning for the next column

Looking at the list of search options for Computers for example (not complete example):
- common => Common
- 1 => [...]
- 86 => [...]
- ...
- operatingsystem => Operating System
- 45 => [...] // This is the name field

For the OS name column (search option ID 45), it iterates the options until it reaches key 45, then goes backwards until it reaches `operatingsystem`. It then assigns the group name "Operating System" to the column for search option 45.

For the display preference list, a similar process would need to be done.

**Review by trasher — APPROVED** [review]

File: src/DisplayPreference.php [hunk 1]
```diff
@@ -466,6 +467,38 @@ class='btn'>";
...
+     * The first key of string type we encouter, is our item's group name.
+     *
+     * @param array $searchopt
+     * @param int   $searchoptkey
+     *
+     * @return string Return the name of the group or an empty string.
+     *
+     * @since 10.0.8
+     */
+    private static function nameOfGroupForItemInSearchopt($searchopt, $searchoptkey)
```
Comment (cedric-anne) [inline src/DisplayPreference.php:488]: 1. `static` methods should avoided when it is possible.
2. Parameters types and return values types should be defined when it does not introduce a breaking change.
```suggestion
    private function nameOfGroupForItemInSearchopt(array $search_options, int $search_option_key): string
```

File: src/DisplayPreference.php [hunk 2]
```diff
@@ -559,6 +592,7 @@ public function showFormGlobal($target, $itemtype)
                     && isset($searchopt[$data["num"]])
                 ) {
                     echo "<tr><td>";
+                    echo self::nameOfGroupForItemInSearchopt($searchopt, $data["num"]);
```
Comment (orthagh) [inline src/DisplayPreference.php:595]:
```suggestion
                    echo $this->nameOfGroupForItemInSearchopt($searchopt, $data["num"]);
```

File: src/DisplayPreference.php [hunk 3]
```diff
@@ -410,6 +410,7 @@ class='btn'>";
                     if ((!in_array($data["num"], $fixed_columns)) && isset($searchopt[$data["num"]])) {
                         echo "<tr>";
                         echo "<td>";
+                        echo self::nameOfGroupForItemInSearchopt($searchopt, $data["num"]);
```
Comment (cedric-anne) [inline src/DisplayPreference.php:413]:
```suggestion
                        echo $this->nameOfGroupForItemInSearchopt($searchopt, $data["num"]);
```

File: src/DisplayPreference.php [hunk 4]
```diff
@@ -466,6 +467,38 @@ class='btn'>";
...
+     */
+    private function nameOfGroupForItemInSearchopt(array $search_options, int $search_option_key): string
+    {
+        $searchoptkeys = array_keys($searchopt);
+
+        for ($key = array_search($searchoptkey, $searchoptkeys) - 1; $key > 0; $key--) {
+            if (is_string($searchoptkeys[$key])) {
+                return $searchopt[$searchoptkeys[$key]]['name'] . " - ";
+            }
+        }
```
Comment (cedric-anne) [inline src/DisplayPreference.php:496]:
```suggestion
        $search_options_keys = array_keys($search_options);

        for ($key = array_search($search_option_key, $search_options_keys) - 1; $key > 0; $key--) {
            if (is_string($search_options_keys[$key])) {
                return $search_options[$search_options_keys[$key]]['name'] . " - ";
            }
        }
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by orthagh — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
Thanks for the contribution.

**Pull request merged**